daemon/src/collab_settlement/protocol.rs

```diff
@@ -0,0 +1,229 @@
+                unsigned_tx: unsigned_tx.clone(),
+                error: anyhow!("failed to received msg3"),
+            });
+        }
+    };
+
+    let collab_settlement_tx =
+        match collab_settlement_tx.recv_counterparty_signature(msg3.listener_signature) {
+            Ok(collab_settlement_tx) => collab_settlement_tx,
+            Err(error) => {
+                return Err(DialerFailed::AfterSendingSignature {
+                    unsigned_tx: unsigned_tx.clone(),
+                    error,
+                });
+            }
+        };
+
+    let settlement =
+        collab_settlement_tx
+            .finalize()
+            .map_err(|e| DialerFailed::AfterSendingSignature {
+                unsigned_tx: unsigned_tx.clone(),
+                error: e,
+            })?;
+
+    Ok(settlement)
+}
+
+#[derive(Debug)]
+pub enum DialerFailed {
+    Rejected,
+    AfterSendingSignature {
+        unsigned_tx: Transaction,
+        error: anyhow::Error,
+    },
+    BeforeSendingSignature {
+        source: anyhow::Error,
+    },
+}
+
+impl From<anyhow::Error> for DialerFailed {
+    fn from(source: anyhow::Error) -> Self {
+        Self::BeforeSendingSignature { source }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub enum DialerMessage {
+    Propose(Propose),
+    Msg2(Msg2),
+}
+
+impl DialerMessage {
+    pub fn into_propose(self) -> Result<Propose> {
+        match self {
+            DialerMessage::Propose(msg0) => Ok(msg0),
+            DialerMessage::Msg2(_) => Err(anyhow!("Expected Msg0 but got Msg2")),
+        }
+    }
+
+    pub fn into_msg2(self) -> Result<Msg2> {
+        match self {
+            DialerMessage::Msg2(msg2) => Ok(msg2),
+            DialerMessage::Propose(_) => Err(anyhow!("Expected Msg2 but got Msg0")),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Serialize, Deserialize)]
+pub enum ListenerMessage {
+    Msg1(Decision),
+    Msg3(Msg3),
+}
+
+impl ListenerMessage {
+    pub fn into_msg1(self) -> Result<Decision> {
+        match self {
+            ListenerMessage::Msg1(msg1) => Ok(msg1),
+            ListenerMessage::Msg3(_) => Err(anyhow!("Expected Msg1 but got Msg3")),
+        }
+    }
+
+    pub fn into_msg3(self) -> Result<Msg3> {
+        match self {
+            ListenerMessage::Msg3(msg3) => Ok(msg3),
+            ListenerMessage::Msg1(_) => Err(anyhow!("Expected Msg3 but got Msg1")),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Propose {
+    pub id: OrderId,
+    pub price: Price,
+    /// The transaction that is being proposed to close the protocol.
+    ///
+    /// Sending the full transaction allows the listening side to verify, how exactly the dialing
+    /// side wants to perform collaborative settlement.
+    #[serde(with = "hex_transaction")]
+    pub unsigned_tx: Transaction,
+}
+
+#[derive(Clone, Copy, Serialize, Deserialize)]
+pub enum Decision {
+    Accept,
+    Reject,
+}
+
+#[derive(Clone, Copy, Serialize, Deserialize)]
+pub struct Msg2 {
```

Comment:
I find it a bit odd to have `Msg2` and `Msg3` but no `Msg1` because it is the decision.

Maybe we should just rename `Msg2` and `Msg3` to what they are (i.e. `DialerSignature` and `ListenerSignature`).

Reply:
Great point. the numeration was indeed off. I have added a small commit that cleans this up (including comments and variable names).